Upgrade worker

karaxuna opened the pull request:
***

**Please remember to write tests for your changes**. We aim to automatically
deploy `master` to production, and we can't safely do this without a solid
test suite!

Ereski commented on apps/server/lib/http/routes.ts:
```diff
@@ -172,7 +175,7 @@ export const attachRoutes = (
 				return response.status(200).json({
 					error: false,
-					data: _.omit(results.data, ['links']),
+					data: _.omit(results.data as any, ['links']),
```
There's almost no valid reason for an `as any` cast. Is it really necessary here?

Ereski commented on apps/server/lib/http/routes.ts:
```diff
@@ -348,7 +351,7 @@ export const attachRoutes = (
 
 			return response.status(200).json({
-				access_token: createSessionResult.data.id,
+				access_token: (createSessionResult.data as any).id,
```
Another `as any` cast

karaxuna replied:
@Ereski `waitResults` should be generic and accept return type right?

Ereski replied:
@karaxuna which `waitResults`?

karaxuna replied:
@Ereski https://github.com/product-os/jellyfish/blob/master/apps/server/lib/http/routes.ts#L341

Ereski replied:
How can you be sure that it is impossible for `createSessionResult.data` to be of any type other than object? https://github.com/product-os/jellyfish-types/blob/master/lib/queue/contracts/execute.ts#L16-L24

karaxuna replied:
Not sure how it works internally. It may return value with one of these types, in which case we don't handle all of them properly and not only typing is a problem, otherwise it's just typing problem and we should tell `waitResults` what type we expect to be returned.

Ereski replied:
@karaxuna sounds like this piece of code needs some extra error checking instead of a blanket `as any`

Ereski commented on apps/server/package.json:
```diff
@@ -25,7 +25,7 @@
   "dependencies": {
```
Should also drop `jellyfish-plugin-base` from here

karaxuna replied:
Dropped plugin-base, but there are some errors because of outdated plugin versions. I'll fix it when all plugins are updated.

Ereski commented on package.json:
```diff
@@ -84,7 +84,7 @@
     "@balena/jellyfish-chat-widget": "^12.5.102",
```
Drop `jellyfish-plugin-base` from here too

Ereski reviewed (commented):
@karaxuna Lots of small things slipped by

karaxuna committed:
Upgrade worker to v15

Change-type: minor

Ereski commented on apps/server/lib/http/facades/auth.ts:
```diff
@@ -45,23 +47,23 @@ export class AuthFacade extends QueryFacade {
+				},
+				ipAddress,
+			)
+		)[0] as Contract | null;
```
`queryAPI` is an internal method. Why isn't it returning the right type?

karaxuna replied:
```typescript
const a: Contract[] = [];
const b = a[0];
```

Type of `b` is `Contract`, but I expected `Contract | undefined`. This is the reason why removing cast causes an error.

Ereski replied:
https://stackoverflow.com/questions/50647399/typescript-accessing-an-array-element-does-not-account-for-the-possibility-of-u

Ereski replied:
:thinking:

karaxuna replied:
@Ereski adding  `--noUncheckedIndexedAccess` sounds good for me, what do you think?

Ereski replied:
Sounds good to me too! And in `jellyfish-config` if it's possible too!

karaxuna replied:
@Ereski https://github.com/product-os/jellyfish-config/pull/60

Ereski commented on apps/server/lib/http/facades/query.ts:
```diff
@@ -57,12 +54,7 @@ export class QueryFacade {
 			});
 
-			const data = await this.jellyfish.query(
-				context,
```
To avoid confusing anyone in the future, would be nice to rename everything of type `LogContext` to `logContext` or similar. `context` is too general for such a restricted value and JF has multiple `*Context` types. Not saying it has to be like that tho

Ereski commented on apps/server/lib/plugins.ts:
```diff
@@ -11,24 +10,23 @@ import { GitHubPlugin } from '@balena/jellyfish-plugin-github';
 import { OutreachPlugin } from '@balena/jellyfish-plugin-outreach';
 import { ProductOsPlugin } from '@balena/jellyfish-plugin-product-os';
 import { TypeformPlugin } from '@balena/jellyfish-plugin-typeform';
+import { PluginManager } from '@balena/jellyfish-worker/build/plugin';
```
No deep imports (`jellyfish-types` is an exception). `jellyfish-worker` already exports `PluginManager` from the root

Ereski reviewed (commented):
Based on the diffs it looks good. Just a few comments here and there

karaxuna committed:
Fix types

Ereski committed:
Upgrade most dependencies

Ereski commented:
There are some major version upgrades we gotta/can do. And remove the deprecated `request` package too

Ereski committed:
Fix most type errors in the server

Ereski commented:
@joshbwlng @karaxuna in `server/lib/bootstrap.ts` we need to remove deep imports from the worker and move the `StreamChange` type into the core.

Ereski committed:
Fix imports and add missing deps in the UI

Ereski commented:
@karaxuna your force pushes completely erased the work I did yesterday

karaxuna commented:
> @karaxuna your force pushes completely erased the work I did yesterday

Oh my. What kind of changes are lost? :(

Ereski commented:
dependency updates, addition of missing deps, some types fixes and multiple import fixes. It was quite a bit

Ereski commented:
just need to know what you added to your last pushes to reconciliate our changes @karaxuna

Ereski committed:
Update and add missing deps, and remove deep imports from the worker

Ereski committed:
Remove even more deep imports from server/bootstrap.ts

LucianBuzzo committed:
server: Add more rigorous type checking when looking up token

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

LucianBuzzo committed:
Server: Cast to any to resolve type error

We will follow up with a root cause fix for the type error once the
worker upgrade is complete and the deploy pipeline is unblocked.

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

LucianBuzzo committed:
Server: remove unnecessary unit test of view facade

The view facade is called during e2e testing, which would flag up any
issues. It doesn't contain complicated logic that requires assertion and
doesn't provide a user facing interface. Unit testing it adds
unnecessary overhead.

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

Ereski committed:
Update worker

LucianBuzzo commented:
Now failing with `Server error {"name":"TypeError","message":"jellyfish.defaults is not a function","stack":"TypeError: jellyfish.defaults is not a function\n    at loadCards (/usr/src/jellyfish/apps/server/build/card-loader.js:15:86)\n    at async bootstrap (/usr/src/jellyfish/apps/server/build/bootstrap.js:274:21)"}`

LucianBuzzo committed:
server: Switch to using static `defaults` method in card loader

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

LucianBuzzo commented:
The issue with `.defaults` is resolved, now we need to fix some type errors that are thrown when unit testing the UI.

LucianBuzzo committed:
UI: Improve type safety of LinkModal

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

LucianBuzzo committed:
UI: Improve typing for filter handling in View component

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

Ereski commented:
Tests failing with:

```
JellyfishDatabaseError: sorry, too many clients already
```

LucianBuzzo committed:
UI: Improve types for View lens

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

LucianBuzzo committed:
UI: Fix typings for View unit tests

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

LucianBuzzo committed:
UI: Fix numerous type errors in unit tests

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

LucianBuzzo committed:
UI: Fix type errors in ViewFilters component

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

LucianBuzzo committed:
UI: Fix promise flushing test utility for jest v27

Jest v27 makes a number of breaking changes to `setImmediate`. This
updated moves to a different method of flushing the promise queue.

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

Ereski committed:
Regenerate lockfiles

Ereski committed:
Remove nock from the root package.json as it was causing the linter to error out

karaxuna commented:
@balena-ci retest

Ereski committed:
Fix the login handler using the wrong `Worker` type

LucianBuzzo committed:
Server: Fix incorrect logContext field in action facade

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

Ereski committed:
Drop useless `as any` casting and simplify a call to `kernel.replaceCard`

LucianBuzzo committed:
server: improve type safety of action facade

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

Ereski committed:
Run integration tests before e2e

Ereski committed:
Improve typing of the oauth module and fix call to `Sync.whoami`

Ereski committed:
More typing and fixes around server initialization

Ereski committed:
Update `jellyfish-plugin-balena-api` with an important fix

LucianBuzzo committed:
Upgrade to latest worker version

This update fixes issues with:
- log context wrapping
- circular references in action request errors
- incorrect handling of http failures during ouath flow

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

karaxuna commented:
@balena-ci retest

LucianBuzzo committed:
server: Improve typing of outreach-mock

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

LucianBuzzo committed:
server: Don't reset outreach mock between test cases

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

LucianBuzzo committed:
server: use meaningful usernames in outreach mirror test

Using more descriptive usernames will help debug mirror tests when they
don't work correctly or are flakey.

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

karaxuna committed:
Fix sync.isValidEvent argument order

karaxuna commented:
@balena-ci retest

LucianBuzzo commented:
It's know failing to validate balena-api webhooks. I can't currently replicate the issue at home, so I've made a debug branch for balena-api to test this further https://github.com/product-os/jellyfish-plugin-balena-api/pull/306

karaxuna commented:
@LucianBuzzo I don't see failing tests [here](https://ci.product-os.io/builds/482357). I only see:

```
Could not find commit hash for previous e2e-server dump
```

LucianBuzzo commented:
@karaxuna urgh weird, that means the `jose` test is flakey then?

karaxuna commented:
@LucianBuzzo you mean the test that I fixed with last commit?

LucianBuzzo commented:
@karaxuna Amazing! I didn't even notice you'd pushed that 😂 - do the typings for that call need to be fixed?

LucianBuzzo commented:
Sick it passed! time to rebase and tighten up those types

LucianBuzzo committed:
test: increase commits scanned to 40 when looking for pg dump hash

Change-type: patch
Signed-off-by: Lucian Buzzo <lucian.buzzo@gmail.com>

LucianBuzzo commented:
@balena-ci retest
